File: ios/alier_dev/alier/_PathRegistry.swift
/*
Copyright 2024 Suredesigns Corp.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import Foundation
import UniformTypeIdentifiers
import os

public class _PathRegistry {
    enum PathKind {
    case isDirectory
    case isFile
    case notExist
    }
    /**
     * Directory on App-specific storage used for Alier
     *
     * which may be `"/data/user/0/{app-package}/app_alier"`.
     */
    var frameworkRootDir: URL {
        get { return _alier_dir }
    }

    /**
     * The base URL for accessing files stored in the user's device.
     * The base URL may vary each time the application is launched.
     */
    var _alier_dir: URL {
        get {
            let default_url = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
            return default_url._appending(path: "Shared", directoryHint: .isDirectory)
        }
    }
    
    init(){
        self.getAppDataDir()
    }

    /**
     * Make a directory under the Alier specific directory if it doesn't exist.
     *
     * - Parameters:
     *   - dirname: the primary segment of the target directory
     *   - subdirs: a sequence of segments succeeding the primary segment
     * - Returns: an `URL` representing the target directory
     */
    @discardableResult
    public func mkdir(dirname: String, subdirs: String...) throws -> URL {
        var dir_to_be_created: URL = _alier_dir._appending(path: dirname, directoryHint: .isDirectory)
        var created_dirs: [URL] = []
        let fm = FileManager.default
        do {
            //  create `dir_to_be_created` directory if not exist
            if !isDirectory(at: dir_to_be_created) {
                try fm.createDirectory(at: dir_to_be_created, withIntermediateDirectories: true)
                created_dirs.append(dir_to_be_created)
            }
            //  create subsequent directories as well
            for segment in subdirs {
                //  skip empty segments
                if segment.isEmpty {
                    continue
                }
                
                dir_to_be_created = dir_to_be_created._appending(path: segment, directoryHint: .isDirectory)

                // skip existing directories
                if isDirectory(at: dir_to_be_created) { continue }
                
                try fm.createDirectory(at: dir_to_be_created, withIntermediateDirectories: true)
                created_dirs.append(dir_to_be_created)
            }
        } catch {
            //  remove created directories if mkdir failed.
            for i in created_dirs.indices.reversed() {
                let created_dir = created_dirs[i]
                try? fm.removeItem(at: created_dir)
            }
            //  notify failure to the caller
            throw error
        }
        
        return dir_to_be_created
    }
    
    /**
     Get `/alier/app_res` directory on the App-specific storage.
     */
    public func getAppResDir() -> URL {
        return try! mkdir(dirname: "app_res")
    }
    
    /**
     Get `/alier/data` directory on the App-specific storage.
     */
    func getAppDataDir() -> URL {
        return try! mkdir(dirname: "app_data")
    }
    
    /**
     Get `/alier/app_res/.temp` directory on the App-specific storage.
     */
    func getAppResTempDir() -> URL {
        return try! mkdir(dirname: "app_res", subdirs:".temp")
    }
    
    /**
     Get `/alier/system` directory on the App-specific storage.
     */
    func getSystemDir() -> URL {
        return try! mkdir(dirname: "alier_sys")
    }
    
    /**
     Get `/alier/system/.temp` directory on the App-specific storage.
     */
    func getSystemTempDir() -> URL {
        return try! mkdir(dirname: "alier_sys", subdirs: ".temp")
    }
    
    /**
     * Gets the `__base.html` file path.
     *
     * - Returns: an `URL` pointing to the `__base.html` file path.
     */
    func getBaseHtmlPath() -> URL {
        return getSystemDir()._appending(path: "__base.html", directoryHint: .notDirectory)
    }
    
    /**
     * Normalize the given path.
     * If the given path starts with "/" it is treated as a relative path to the root.
     * If the given path starts with ".." it is treated as a relative path to the parent of the app_res directory.
     *
     * @param path A string representing a path to be calculated a relative path from
     * the root of the app-specific data directory.
     * @return a [String] representing a path relative to the root of the app-specific data directory.
     * @throws [IllegalArgumentException] when the given [path] is outside of the root directory.
     */
    func normalizePathString(path: String) -> String {
        if path.isEmpty { return "" }
        
        let cwd = "/" + getAppResDir().lastPathComponent
        
        // Split the path at "/" and store it in an array
        var path_components = path.components(separatedBy: "/")
        var i = 0
        // Remove "..", ".", and "" from path_components
        while i < path_components.count {
            switch path_components[i] {
            case "..":
                if i >= 1 {
                    let prev = path_components[i - 1]
                    if prev == "/" || prev == ".." {
                        i += 1
                    } else {
                        path_components.remove(at:i)
                        path_components.remove(at:i - 1)
                        i -= 1
                    }
                } else {
                    i += 1
                }
            case ".":
                path_components.remove(at:i)
            case "":
                if i >= 1 {
                    path_components.remove(at: i)
                } else {
                    i += 1
                }
            default:
                i += 1
            }
        }
        // Create the normalized path
        let path_normalized: String
        // If path_components is empty, assign ""
        if path_components.isEmpty {
            path_normalized = ""
        } else {
            if path_components[0] == ".." {
                if cwd == "/" {
                    try! _PathRegistry.erroThrow(UniqueErrors.IllegalArgumentException("Given path is outside of the root: \(path)"))
                }
                var path_prefix = cwd
                while !path_components.isEmpty && path_components[0] == ".." {
                    if path_prefix.isEmpty {
                        try! _PathRegistry.erroThrow(UniqueErrors.IllegalArgumentException("Given path is outside of the root: \(path)"))
                        path_components.remove(at:0)
                        path_prefix = String(path_prefix[path_prefix.startIndex..<path_prefix.lastIndex(of: "/")!])
                    }
                    path_components.insert(path_prefix, at: 0)
                }
                path_normalized = path_components.joined(separator: "/")
            } else {
                let path_joined = path_components.joined(separator: "/")
                if path_joined.isEmpty {
                    path_normalized = "/"
                } else {
                    path_normalized = path_joined
                }
            }
        }
        return path_normalized
    }
    
    func getFilePath(path: String) -> URL? {
        let path_normalized = normalizePathString(path: path)
        let file_url: URL = if path_normalized.hasPrefix("/") {
            _alier_dir._appending(path: path_normalized, directoryHint: .notDirectory)
        } else {
            getAppResDir()._appending(path: path_normalized, directoryHint: .notDirectory)
        }
        
        let file_path: String = file_url._path(percentEncoded: false)
        if !isFile(atPath: file_path) {
            return nil
        }
        return file_url
    }
    
    /// Gets framework's bundle object.
    ///
    /// - Returns: `Bundle` of framework' resources.
    func getSystemBundle() -> Bundle {
        #if SWIFT_PACKAGE
        return Bundle.module
        #else
        return Bundle(for: _PathRegistry.self)
        #endif
    }
    
    /// Gets filepaths of assets included in the given bundle.
    ///
    /// - Parameters:
    ///   - bundle: the target bundle
    ///   - forResource: the target resource name in the given bundle.
    /// - Returns: an array of filepaths of resources in the target bundle if they exist.
    /// - Throws: when the target resource is not a directory, or inaccesible, or broken
    func getBundleFiles(bundle: Bundle, forResource: String?) throws -> [URL]? {
        guard let resource_root: URL = bundle.url(forResource: forResource, withExtension: nil) else {
            return nil
        }

        let fm = FileManager.default

        let property_keys = Set<URLResourceKey>([.isRegularFileKey])
        guard let resource_enumarator = fm.enumerator(at: resource_root, includingPropertiesForKeys: Array(property_keys)) else {
            return nil
        }
    
        var resources: [URL] = []
        for case let resource as URL in resource_enumarator {
            guard let resource_values = try? resource.resourceValues(forKeys: property_keys),
                  let is_regular_file: Bool = resource_values.isRegularFile
            else {
                continue
            }
            
            if is_regular_file {
                resources.append(resource.resolvingSymlinksInPath())
            }
        }
        
        return resources
    }
    
    
    /// Gets assets bundled with application and framework.
    ///
    ///- Returns: an array of URLs of all assets.
    func getAssets() -> ([URL], [URL]) {
        let system_assets = try! getBundleFiles(bundle: getSystemBundle(), forResource: "system")!
        let app_assets = try! getBundleFiles(bundle: Bundle.main, forResource: "app_res")!
        
        return (system_assets, app_assets)
    }

    enum UniqueErrors: Error {
        case IllegalArgumentException(String)
        case nilValueError
        case otherError(String)
    }
    
    static func erroThrow(_ error: UniqueErrors) throws {
        switch error {
        case .IllegalArgumentException:
            throw error
        case .nilValueError:
            throw error
        case .otherError:
            throw error
        }
    
    }
    
    func kindOf(atPath path: String) -> PathKind {
        var is_dir: ObjCBool = false
        let exists = FileManager.default.fileExists(atPath: path, isDirectory: &is_dir)

        return if is_dir.boolValue {
            .isDirectory
        } else if exists {
            .isFile
        } else {
            .notExist
        }
    }

    func kindOf(at fileurl: URL) -> PathKind {
        return kindOf(atPath: fileurl._path(percentEncoded: false))
    }

    func isFile(atPath path: String) -> Bool {
        kindOf(atPath: path) == .isFile
    }

    func isFile(at fileurl: URL) -> Bool {
        kindOf(at: fileurl) == .isFile
    }
    
    func isDirectory(atPath path: String) -> Bool {
        kindOf(atPath: path) == .isDirectory
    }
    
    func isDirectory(at fileurl: URL) -> Bool {
        kindOf(at: fileurl) == .isDirectory
    }

}

enum UniqueErrors: Error {
    case IllegalArgumentException(String)
    case nilValueError
    case IOException(String)
}

func erroThrow(_ error: UniqueErrors) throws {
    switch error {
    case .IllegalArgumentException:
        throw error
    case .nilValueError:
        throw error
    case .IOException:
        throw error
    }
}

File: android/alier_dev/alier/src/main/java/com/suredesigns/alier/_PathRegistry.kt
/*
Copyright 2024 Suredesigns Corp.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package com.suredesigns.alier

import android.content.Context
import java.io.File
import java.io.IOException
import kotlin.IllegalArgumentException

/**
 * This class provides functions accessing files / directories on the App-specific storage.
 *
 * To operate a file or a directory gotten from this registry, use [_FileOperation] class.
 */
class _PathRegistry(
    private val _application_context: Context
) {
    init{
        this.getAppDataDir()
    }

    /**
     * Directory on App-specific storage used for Alier
     *
     * which may be `"/data/user/0/{app-package}/app_alier"`.
     */
    val frameworkRootDir: File
        get() = _application_context.getDir("alier", Context.MODE_PRIVATE)

    /**
     * Makes a sub-directory under the Alier specific directory if it doesn't exist,
     * and then return the requested path as a File object.
     *
     * @param dirname a string representing a directory name relative to the directory on
     * the App-specific storage used for the framework.
     * Note that if an absolute path is given as this argument,
     * then it is converted to the relative path from the root directory of the framework.
     * This
     * @param subdirs a sequence of sub directories to be made if not exist
     *
     * @return File object having the given directory path.
     * @throws IOException when failing to create a directory.
     * @throws IllegalArgumentException when passing a directory path having different root from
     * the directory used for the framework on the App-specific storage.
     */
    fun mkdir(dirname: String, vararg subdirs: String): File {
        val sep = File.separator
        val basedir = frameworkRootDir.path
        val dirname_rel = if (File(dirname).isAbsolute) {
            dirname.removePrefix(if (dirname.startsWith(basedir)) { basedir + sep } else { sep })
        } else {
            dirname
        }
        val target_dir = File(
            """$basedir$sep$dirname_rel$sep${subdirs.joinToString(sep)}"""
        )
        if (target_dir.isDirectory) {
            return target_dir
        }
        val ls_dir = mutableListOf(basedir, dirname_rel, *subdirs)
        val sb_base_dir = StringBuilder()
        while (ls_dir.isNotEmpty()) {
            val dir = ls_dir.removeAt(0)
            if (dir.contains(sep)) {
                ls_dir.addAll(0, dir.split(sep))
                continue
            }
            sb_base_dir.append(dir, sep)
            if (!File(sb_base_dir.toString()).exists()) {
                break
            }
        }
        ls_dir.add("") // add a sentinel used while-loop below
        val ls_dir_created = mutableListOf<File>()
        try {
            val sb_sub_dir = StringBuilder(sb_base_dir)
            while (ls_dir.isNotEmpty()) { // This is done at least once because ls_dir has a sentinel.
                val file = File(sb_sub_dir.toString())
                if (!file.isDirectory && !file.isFile) {
                    if (!file.mkdir()) {
                        throw IOException("Creating specified directory was failed: $file")
                    }
                }
                val next = ls_dir.removeAt(0)
                ls_dir_created.add(file)
                sb_sub_dir.append(next, sep)
            }
        } catch (e: IOException) {
            for (file in ls_dir_created) {
                if (!file.delete()) {
                    throw IOException("Deletion of disused directory was failed: $file", e)
                }
            }
            throw e
        }
        return target_dir
    }

    /**
     * Gets `/app_alier/app_res` directory on the App-specific storage.
     *
     * @return [File] the directory stored application components.
     * @see getAppDataDir
     * @see getAppResTempDir
     * @see getSystemDir
     * @see getSystemTempDir
     */
    fun getAppResDir(): File {
        return mkdir("app_res")
    }

    /**
     * Gets `/app_alier/app_data` directory on the App-specific storage.
     *
     * @return [File] the directory stored data managed by the framework user.
     * @see getAppResDir
     * @see getAppResTempDir
     * @see getSystemDir
     * @see getSystemTempDir
     */
    @Suppress("MemberVisibilityCanBePrivate")
    fun getAppDataDir(): File {
        return mkdir("app_data")
    }

    /**
     * Gets `/app_alier/app_res/.temp` directory on the App-specific storage.
     *
     * Note that files in the temp directory are NOT treated as cache files.
     * So you should take responsibility for free disk space up by deleting those files manually,
     * if you created files in the temp directory.
     *
     * @return [File] the directory stored application components.
     * @see getAppResDir
     * @see getAppDataDir
     * @see getSystemDir
     * @see getSystemTempDir
     */
    @Suppress("MemberVisibilityCanBePrivate")
    fun getAppResTempDir(): File {
        return mkdir("app_res", ".temp")
    }

    /**
     * Gets `/app_alier/system` directory on the App-specific storage.
     *
     * @return [File] the directory stored system components of Alier.
     * @see getAppResDir
     * @see getAppDataDir
     * @see getAppResTempDir
     * @see getSystemTempDir
     */
    fun getSystemDir(): File {
        return mkdir("alier_sys")
    }

    /**
     * Gets `/app_alier/system/.temp` directory on the App-specific storage.
     *
     * Note that files in the temp directory are NOT treated as cache files.
     * So you should take responsibility for free disk space up by deleting those files manually,
     * if you created files in the temp directory.
     *
     * @return [File] the directory stored temporary files related to system components of Alier.
     * @see getAppResDir
     * @see getAppDataDir
     * @see getAppResTempDir
     * @see getSystemDir
     */
    @Suppress("MemberVisibilityCanBePrivate")
    fun getSystemTempDir(): File {
        return mkdir("alier_sys", ".temp")
    }

    /**
     * Gets `__base.html` file path.
     *
     * @return [File]
     * the `__base.html` file path.
     */
    fun getBaseHtmlPath(): File = File(getSystemDir(), "__base.html")

    /**
     * Normalize the given path.
     * If the given path starts with "/" it is treated as a relative path to the root.
     * If the given path starts with ".." it is treated as a relative path to the parent of the app_res directory.
     *
     * @param path A string representing a path to be calculated a relative path from
     * the root of the app-specific data directory.
     * @return a [String] representing a path relative to the root of the app-specific data directory.
     * @throws [IllegalArgumentException] when the given [path] is outside of the root directory.
     */
    @Suppress("MemberVisibilityCanBePrivate")
    fun normalizePathString(path: String): String {
        if (path.isEmpty()) {
            return ""
        }
        val root = frameworkRootDir.absolutePath
        val cwd = getAppResDir().absolutePath.removePrefix(root)

        //  split the given path into mutable list of segments by the path delimiter.
        val path_segments = path.split(File.separator).toMutableList()
        var i = 0
        //  remove dot-segments and empty segments from the list of segments.
        while (i < path_segments.size) {
            when(path_segments[i]) {
                ".." -> {
                    val prev = if (i >= 1) { path_segments[i - 1] } else { null }
                    if (prev.isNullOrEmpty() || prev == "..") {
                        //  double-dots can be removed if and only if there exists a preceding segment and
                        //  it is neither a dots segment nor an empty segment.
                        i++
                    } else {
                        //  remove a double-dots
                        //  path_segments = foo/BAR/../qux
                        //                          ^^  (i = 2)
                        path_segments.removeAt(i)
                        //                = foo/BAR/qux # double-dot ".." removed
                        //                          ^^^ (i = 2)
                        path_segments.removeAt(i - 1)
                        //                = foo/qux     # preceding segment "BAR" removed
                        //                          ^^^ (i = 2)
                        i--
                        //                = foo/qux     # index moved to the succeeding segment
                        //                      ^^^     (i = 1)
                    }
                }
                "." -> {
                    //  remove a single-dot
                    //  path_segments = foo/bar/./qux
                    //                          ^   (i = 2)
                    path_segments.removeAt(i)
                    //  path_segments = foo/bar/qux # single-dot "." removed
                    //                          ^^^ (i = 2)
                }
                "" -> {
                    if (i >= 1) {
                        //  remove an empty segment
                        //  path_segments = /foo/bar/   /qux
                        //                           ^^^    (i = 2)
                        path_segments.removeAt(i)
                        //  path_segments = /foo/bar/qux
                        //                           ^^^    (i = 2)
                    } else {
                        //  keep the leading empty segment to distinguish absolute paths
                        //  from relative paths.
                        //  path_segments =    /foo/bar/baz
                        //                  ^^^     (i = 0)
                        i++
                        //  path_segments =    /foo/bar/baz
                        //                      ^^^ (i = 1)
                    }
                }
                else -> {
                    i++
                }
            }
        }
        val path_normalized = if (path_segments.isEmpty()) {
            ""
        } else {
            if (path_segments[0] == "..") {
                if (cwd == File.separator) {
                    throw IllegalArgumentException("Given path is outside of the root: $path")
                }
                var path_prefix = cwd
                while (path_segments.isNotEmpty() && path_segments[0] == "..") {
                    if (path_prefix == "") {
                        throw IllegalArgumentException("Given path is outside of the root: $path")
                    }
                    path_segments.removeAt(0)
                    path_prefix = path_prefix.substring(0, path_prefix.lastIndexOf(File.separator))
                }
                path_segments.add(0, path_prefix)
                path_segments.joinToString(File.separator)
            }
            val path_joined = path_segments.joinToString(File.separator)
            if (path_joined == "") { File.separator } else { path_joined }  //  = path_normalized
        }
        return path_normalized
    }

    /**
     * Gets a [File] instance from the given path.
     *
     * @param path A string representing a path.
     * @return a [File] instance representing the given [path].
     * @throws [IllegalArgumentException] when the given [path] is outside of the root directory.
     * @throws [IllegalArgumentException] when a file does not exist on the given [path].
     */
    fun getFilePath(path: String): File? {
        val path_normalized = normalizePathString(path)

        val file = if (path_normalized.startsWith(prefix = File.separator)) {
            File(frameworkRootDir, path_normalized.removePrefix(File.separator))
        } else {
            File(getAppResDir(), path_normalized)
        }

        return if (file.isFile) {
            file
        } else {
            null
        }
    }

    /**
     * Gets path strings of asset files as an array.
     * @return an array of strings representing asset file paths.
     */
    fun getAssetFiles(): Array<String> {
        val assets = _application_context.resources.assets
        val asset_files = assets.list("")?.toMutableList() ?: mutableListOf<String>()
        var i = 0
        while (i < asset_files.size) {
            try {
                // Test whether the given name represents a file or not.
                // if asset_files[i] is not a file, open() will throw IOException.
                assets.open(asset_files[i]).close()
                i++
            } catch (e: IOException) {
                val dir = asset_files.removeAt(i)
                val sub_assets = assets.list(dir) ?: arrayOf()
                for (index in sub_assets.indices) {
                    sub_assets[index] = "$dir${File.separator}${sub_assets[index]}"
                }
                asset_files.addAll(sub_assets)
            }
        }
        return asset_files.toTypedArray()
    }
}